### ISSUE TITLE
Windows Event Log input plugin doesn't pickup correct Event Log.

### ISSUE THREAD

**Issue by rslowe** [fluent/fluent-bit#3383]
## Bug Report

**Describe the bug**
Attempting to use the winlog input plugin with a "Complex" (aka something like "Microsoft-Windows-Time-Service/Operational") name fails to retrieve that log and instead transparently pulls up the "Application" event log instead.

**To Reproduce**
- Open Event Viewer (so you can either verify the log presented by fluent-bit, or use an alternative channel to test)
- Start fluent-bit with the following configuration. (In theory, everyone who is on Windows 10 should have this Time-Service log, if not, use any log in "Applications and Services\Microsoft" in the event viewer, replacing the channel name in the config with the "Full Name" of the log in the log's properties.)
    ```
    [SERVICE]
        Flush           5
        Daemon          yes
        Log_Level       debug
        
    [INPUT]
        Name            winlog
        Channels        Microsoft-Windows-Time-Service/Operational
        Interval_Sec    1
        
    [OUTPUT]
        Name            stdout
    ```
- Observe that the logs outputted are actually the event logs from Application (Windows Logs\Application in Event Viewer)
- Observe that the logs in the pathed log are different (In this case, use Applications and Services\Microsoft\Time-Service\Operational.)

I have tried running this in Admin Mode (Ruling out the cause of transparent Permission Denied), that does not solve the issue.
 
**Expected behavior**
The actual event logs are taken out and printed to stdout. In this case, it should just be mostly events relating to the Time Service (W32TimeSvc)

**Screenshots**
I'm attempting to get this log:
<img width="785" alt="Screenshot 2021-04-19 113102" src="https://user-images.githubusercontent.com/72294336/115285841-391b5b00-a103-11eb-9155-f6e033c6b150.png">

But I'm getting this as output:
<img width="946" alt="Screenshot 2021-04-19 113157" src="https://user-images.githubusercontent.com/72294336/115285879-43d5f000-a103-11eb-986e-75756b707840.png">

This log seems to correspond to this event in Applications (which is the wrong one):
<img width="771" alt="Screenshot 2021-04-19 113324" src="https://user-images.githubusercontent.com/72294336/115285946-5c460a80-a103-11eb-87ca-2b5c5a6807ec.png">

**Your Environment**

* Version used: 1.7.3
* Configuration: 
    ```
    [SERVICE]
        Flush           5
        Daemon          yes
        Log_Level       debug
        
    [INPUT]
        Name            winlog
        Channels        Microsoft-Windows-Time-Service/Operational
        Interval_Sec    1
        
    [OUTPUT]
        Name            stdout
    ```
* Environment name and version (e.g. Kubernetes? What version?): Bare Metal Laptop
* Operating System and version: Windows 10, 64-bit, Version 2004 (19041.867)
* Filters and plugins: winlog input, stdout output.

**Additional context**
I'm attempting to gather windows event logs to eventually get them to a file. The file part works so I'm not including that in my config. However, fundamentally it seems that fluent-bit is not gathering the correct logs (for this event log at least).

I've noticed in https://github.com/fluent/fluent-plugin-windows-eventlog, that there is a "windows_eventlog2" type for fluentd, which makes me think that there's a newer API for windows to gather windows event logs. 

I was wondering if there's something equivalent over here on the fluent-bit side, and if so, something to point me in the right direction to it.

**Comment by rslowe** [comment]
Uping to avoid stale. Anyone else encounter this issue?

**Comment by dhwest14** [comment]
I'm having the same issue. Events that are showing in the Windows Defender/Operational and Sysmon/Operational channels are not getting passed by fluentbit, but instead events are taken from the Application events log file and forwarded that don't exist in the original channel.

**Comment by Noly** [comment]
I faced same issue with events from `Channels Microsoft-Windows-Windows Defender/Operational` in Win 2019. Events are capture from Application and without any message; 
<img width="1524" alt="Screen Shot 2021-06-25 at 13 53 38" src="https://user-images.githubusercontent.com/4612272/123421135-c1fbba80-d5bc-11eb-8752-02449f98fff6.png">

**Comment by noly** [comment]
Avoid stale. Hopping for any update from fluent bit team :)

**Comment by noly** [comment]
Avoid stale. Hopping for any update from fluent bit team :)

**Comment by agup006** [comment]
There is a newer windows API and I believe we are still using the old one. We might be able to add a better custom path with the old api - @cosmo0920 could you take a look next week?

**Comment by cosmo0920** [comment]
Short description: Yeah, winlog plugin uses old API to retrieve Windows EventLog. New EventLog channels is not compatible for new APIs which are provided from [`<winevt.h>`](https://docs.microsoft.com/en-us/windows/win32/api/winevt/).
I'll take a look on this week.

**Comment by cosmo0920** [comment]
Hmm..., it's weird. In my Windows laptop, Microsoft-Windows-Time-Service/Operational can be consumed with `winlog` plugin.
But, string inserts does not handled properly to inject Windows EventLog string template.

 
![image](https://user-images.githubusercontent.com/700876/134325506-2613df1e-7d5c-48a5-b41e-cce5b0784866.png)

**Comment by rslowe** [comment]
@cosmo0920, It does appear to look like that, but if you manually look at the Time Service Operational logs in Event Viewer, they do not match at all. (Based on your screenshot, it appears that "Software Protection Platform Service" is in the Time Service Log, which does not make any sense.)

Just to be certain that this still does cause an issue, I've tried this on FluentBit v1.8.7 (the current version as of writing), and can confirm that the issue still exists.

An example line output from Fluent-Bit when using the above config is:
`[132] winlog.0: [1632321564.238334100, {"RecordNumber"=>3688, "TimeGenerated"=>"2021-09-22 07:32:21 -0700", "TimeWritten"=>"2021-09-22 07:32:21 -0700", "EventID"=>1034, "Qualifiers"=>16384, "EventType"=>"Information", "EventCategory"=>0, "Channel"=>"Microsoft-Windows-Time-Service/Operational", "SourceName"=>"Software Protection Platform Service", "ComputerName"=>"Redacted_Computer_Name", "Data"=>"", "Sid"=>"", "Message"=>"", "StringInserts"=>["AAD-BlockAADWorkplaceJoin-Default", "100"]}]`

The actual corresponding output is in the Application log, not the Time Service log, per the Windows Event Viewer.
![image](https://user-images.githubusercontent.com/72294336/134366009-ae69806f-453b-4836-905c-a2ce9af788cc.png)

Meanwhile, Time-Service/Operational logs have events like this, and none of the "Software Protection Platform Service" entries.
![image](https://user-images.githubusercontent.com/72294336/134366409-39918c6e-d9a9-47f7-9f3f-94f4819d97ef.png)

**Comment by cosmo0920** [comment]
Right. fluent-bit's winlog plugin does not implement win32-eventlog equivalent getting event description functionality:
https://github.com/chef/win32-eventlog/blob/4ee4c33665e0774963539e36e890108e45466f9e/lib/win32/eventlog.rb#L854

So, string inserts values are only obtained. 
They are used for injecting string template.

String template is used by reducing windows eventlog's event size like as:

```
Leap second configuration:
Enabled: {0}
Current: {1}
Current Offset from UTC(Seconds): {2}
Runtime state consistent with settings: {3}
Newest Leap Seconds List(Local): {4}
```

And string inserts are:
```
["1", "0 (Local)", "0", "1", ""]
```

Finally, they will be concatenated and be Event Viewer like appearance.

Currently, winlog just retrieved the above string inserts values.
For appearance, we must obtain string template, string inserts and restore original messages under `HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Services\\EventLog\\` registry keys saved EventMessageFile.

**Comment by cosmo0920** [comment]
I'm continuing further investigation for this issue.
Currently, I'd found that classical channels are retrievable for ReadEventLog and Registry related APIs.
But, for the new event channels that are managed under applications and services branch, current winlog plugin is not able to handle them.
We're planning to provide a new plugin which uses `winevt.h` provided APIs.

**Comment by cosmo0920** [comment]
FYI: Event Viewer seems to be built with C# and System.Diagnostics.EventLog class. This class' backend also uses `winevt.h` API. 
So, Event Viewer works nicely.

**Comment by cosmo0920** [comment]
I sent a PR to handle the new Windows EventLog plugin:
https://github.com/fluent/fluent-bit/pull/4179

The old winlog plugin does not handle Applications and Services Logs.
This is known issue. The new plugin will be able to fix it.

**Comment by ixniz** [comment]
> I also ran into this immediately on my first day trying out fluent bit.

+1
I also ran into this immediately, leading me to find this issue suspecting the "new" event logs weren't handled correctly.

**Comment by cosmo0920** [comment]
> I also ran into this immediately, leading me to find this issue suspecting the "new" event logs weren't handled correctly.

Yep. Classic Windows EventLog channels are handled correctly with the old one.
But, `Microsoft-Windows-PowerShell/Operational` is not one of the classic EventLog channels but new one.
"new" EventLog channels are stored with `evtx` format that is not handled correctly by `winlog` plugin.

**Comment by TrevorBergman** [comment]
Has there been any movement on this topic? I need to forward multiple logs from the "Applications and Services Logs" branch. Unfortunately, I seem to be running into the very issue Cosmo0920 calls out in his post on Dec 29 (no support in Fluentbit for this 'newer' branch). Hoping that something might have been done and if nothing else, get a status on whether this is something still being addressed from the dev perspective.

**Comment by cosmo0920** [comment]
Fluent Bit. V1.9.0.rc1 starts to support this newer Windows EventLog plugin (winevtlog).

**Comment by TrevorBergman** [comment]
Thanks Cosmo! I appreciate the update. Would you have any insight into when this new release would be released? Anything you can provide would be appreciated. We are trying to make a decision and knowing if we are days, weeks, or months away from this release would be invaluable.

**Comment by cosmo0920** [comment]
PR https://github.com/fluent/fluent-bit/pull/4179 is already included in fluent-bit v1.9.0.
@patrick-stephens @edsiper Can we close this issue?

**Comment by agup006** [comment]
Plus 1 to closing
